gen/emit/EmitJulia.cpp

```diff
@@ -548,27 +592,194 @@ struct EmitJulia : public Emitter
+        // can define foo.bar.baz
+        set<string> moduleJlSubmodules;
+        moduleJlSubmodules.insert(packName);
+
+        FILE *moduleJlFp = nullptr;
+        if (havePackage) {
+            // If we are inside a package, then we need to generate a single
+            // Julia file containing the top-level module and any submodules
+
+            // For a type foo.bar.baz.t1, we will put the module file in
+            // foo.jl
+            vector<string> moduleJlFnameParts;
+            moduleJlFnameParts.push_back(packageDirPrefix);
+            moduleJlFnameParts.push_back(pdname + ".jl");
+            string moduleJlFname = StringUtil::join(moduleJlFnameParts, '/');
+            if (moduleJlFp) {
+                fclose(moduleJlFp);
+                moduleJlFp = nullptr;
+            }
+            if (FileUtil::exists(moduleJlFname)) {
+                // If the module exists already, then we need to parse it and
+                // extract any existing sub-modules and types contained in the
+                // module. We'll include those entries in the final generated
+                // module file.
+                moduleJlFp = fopen(moduleJlFname.c_str(), "r");
+                if (!moduleJlFp) {
+                    perror("fopen");
+                    return -1;
+                }
+                // Read chunks of the existing file line by line
+                while(!feof(moduleJlFp)) {
+                    char buf[4096];
+                    memset(buf, 0, sizeof(buf));
+                    char *result = fgets(buf, sizeof(buf)-1, moduleJlFp);
+                    if (!result) break;
+
+                    auto words = StringUtil::split(StringUtil::strip(buf), ' ');
+                    if (words.size() >= 2 && words[0] == "import") {
+                        // If this line matches "import foo", then store "foo" in
+                        // the set of imports
+                        moduleJlImports.insert(words[1]);
+                    } else if (words.size() >= 6 &&
+                               words[0] == "@eval" &&
+                               words[2] == "module" &&
+                               words[4] == ";" &&
+                               words[5] == "end") {
+                        // Otherwise, if the line matches:
+                        // @eval foo module bar.baz ; end
+                        // then store foo.bar.baz in the submodules set
+                        moduleJlSubmodules.insert(words[1] + "." + words[3]);
+                    }
+                }
+                fclose(moduleJlFp);
+                moduleJlFp = nullptr;
+            }
+            // Regardless of whether the module existed, we'll create a new file
+            moduleJlFp = fopen(moduleJlFname.c_str(), "w");
+            if (!moduleJlFp) {
+                perror("fopen");
+                return -1;
+            }
+            fprintf(moduleJlFp, "\"\"\"ZCM package %s.jl file\n"
```

Comment:
prefer use of "emit" when writing files in zcmgen (seems like this class should just be the emitter for the module file, right?)

Reply:
Ok, that makes sense. I'm just following the example from EmitPython.cpp.

Reply:
Ah yes, I haven't actually looked at the EmitPython code in a long time (it might even be unchanged since the fork from LCM). Maybe we'll be able to take what we do in the Julia file and go update the Python one when this is all merged.